File: cses/bookShop.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {

    int N, X; cin >> N >> X;
    vector<int> price(N), pages(N);
    for (auto& p : price) cin >> p;
    for (auto& p : pages) cin >> p;
    vector<vector<int>> dp(2, vector<int>(X+1));

    for (int r = 1; r <= N; ++r) {
        int t = r % 2;
        int w = price[r-1];
        int p = pages[r-1];
        for (int c = 1; c <= X; ++c) {
            dp[t][c] = dp[!t][c];
            if (c >= w)
                dp[t][c] = max(dp[t][c], p + dp[!t][c-w]);
        }
    }
    cout << dp[N%2][X] << endl;
}

File: cses/book_shop.py
N, X = map(int, input().split())
price = list(map(int, input().split()))
pages = list(map(int, input().split()))

ans = 0
dp = [[0] * (X+1) for _ in range(2)]
for r in range(1, N+1):
    w = price[r-1]
    p = pages[r-1]
    t = r % 2
    for c in range(1, X+1):
        dp[t][c] = dp[not t][c]
        if c >= w:
            dp[t][c] = max(dp[t][c], p + dp[not t][c-w])

print(dp[N%2][-1])
